File: src/main/java/net/jgp/books/spark/ch12/lab910_flat_json_invoice/FlattenJsonApp.java
package net.jgp.books.spark.ch12.lab910_flat_json_invoice;

import static org.apache.spark.sql.functions.explode;

import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.SparkSession;
import org.apache.spark.sql.types.StructField;
import org.apache.spark.sql.types.StructType;

/**
 * Automatic flattening of JSON structure in Spark.
 * 
 * @author jgp
 */
public class FlattenJsonApp {
  public static final String ARRAY_TYPE = "Array";
  public static final String STRUCT_TYPE = "Struc";

  /**
   * main() is your entry point to the application.
   * 
   * @param args
   */
  public static void main(String[] args) {
    FlattenJsonApp app =
        new FlattenJsonApp();
    app.start();
  }

  /**
   * The processing code.
   */
  private void start() {
    // Creates a session on a local master
    SparkSession spark = SparkSession.builder()
        .appName("Automatic flattening of a JSON document")
        .master("local")
        .getOrCreate();

    // Reads a JSON, stores it in a dataframe
    Dataset<Row> invoicesDf = spark.read()
        .format("json")
        .option("multiline", true)
        .load("data/json/nested_array.json");

    // Shows at most 3 rows from the dataframe
    invoicesDf.show(3);
    invoicesDf.printSchema();

    Dataset<Row> flatInvoicesDf = flattenNestedStructure(spark, invoicesDf);
    flatInvoicesDf.show(20, false);
    flatInvoicesDf.printSchema();
  }

  /**
   * Implemented as a public static so you can copy it easily in your utils
   * library.
   * 
   * @param spark
   * @param df
   * @return
   */
  public static Dataset<Row> flattenNestedStructure(
      SparkSession spark,
      Dataset<Row> df) {
    boolean recursion = false;

    Dataset<Row> processedDf = df;
    StructType schema = df.schema();
    StructField fields[] = schema.fields();
    for (StructField field : fields) {
      switch (field.dataType().toString().substring(0, 5)) {
        case ARRAY_TYPE:
          // Explodes array
          processedDf = processedDf
              .withColumnRenamed(field.name(), field.name() + "_tmp");
          processedDf = processedDf.withColumn(
              field.name(),
              explode(processedDf.col(field.name() + "_tmp")));
          processedDf = processedDf.drop(field.name() + "_tmp");
          recursion = true;
          break;
        case STRUCT_TYPE:
          // Mapping
          String ddl[] = field.toDDL().split("`"); // fragile :(
          for (int i = 3; i < ddl.length; i += 2) {
            processedDf = processedDf.withColumn(
                field.name() + "_" + ddl[i],
                processedDf.col(field.name() + "." + ddl[i]));
          }
          processedDf = processedDf
              .drop(field.name());
          recursion = true;
          break;
      }
    }
    if (recursion) {
      processedDf = flattenNestedStructure(spark, processedDf);
    }
    return processedDf;
  }
}

File: src/main/python/lab900_flat_json_invoice/flattenJsonApp.py
"""
 Processing of invoices formatted using the schema.org format.

 @author rambabu.posa
"""
import os
from pyspark.sql import (SparkSession, DataFrame, functions as F)

array_type = "Array"
struct_type = "Struc"

'''
  It works only for this scenario
  To write this as a generic function, please reimplement this using recursion technique 
'''
def flatten_nested_structure(df:DataFrame) -> DataFrame:
    array_cols = [c[0] for c in df.dtypes if c[1][:5] == 'array']
    new_df = df.select('*', F.explode(*array_cols))

    struct_cols = [c[0] for c in new_df.dtypes if c[1][:6] == 'struct']
    new_df2 = new_df.select(*[c + ".*" for c in struct_cols])

    array_cols2 = [c[0] for c in new_df2.dtypes if c[1][:5] == 'array']
    flat_df = new_df2.select('*', F.explode(*array_cols2)).drop(*array_cols2) \
        .withColumnRenamed("col", "salesByMonth")
    return flat_df

def rename_all_df_columns(df, new_column_names):
    return df.toDF(*new_column_names)

def get_absolute_file_path(path, filename):
    current_dir = os.path.dirname(__file__)
    relative_path = "{}{}".format(path, filename)
    absolute_file_path = os.path.join(current_dir, relative_path)
    return absolute_file_path


def main(spark, path, filename):
    absolute_file_path = get_absolute_file_path(path, filename)

    # Reads a JSON, stores it in a dataframe
    invoices_df = spark.read.format("json") \
        .option("multiline", True) \
        .load(absolute_file_path)

    # Shows at most 3 rows from the dataframe
    invoices_df.show(3)
    invoices_df.printSchema()

    flat_invoices_df = flatten_nested_structure(invoices_df)

    new_columns = ['author_city','author_country','author_name','author_state',
                   'publisher_city','publisher_country','publisher_name','publisher_state',
                   'books_title','books_salesByMonth'
                   ]
    flat_invoices_df = rename_all_df_columns(flat_invoices_df, new_columns) \
        .withColumn('date', F.lit('2019-10-05'))

    flat_invoices_df.show(20, False)
    flat_invoices_df.printSchema()


if __name__ == '__main__':
    # Creates a session on a local master
    spark = SparkSession.builder.appName("Automatic flattening of a JSON document") \
        .master("local[*]").getOrCreate()
    spark.sparkContext.setLogLevel("warn")

    path = '../../../../data/json/'
    filename = 'nested_array.json'

    main(spark, path, filename)
    spark.stop()